### PR TITLE
Do not call handler blocks if they are nil

### PR DESCRIPTION
**Pull request by levinli303**
Calling NSProgress.cancel should not crash when no cancel handler is set.

### CODE REVIEW

**Review by rfm — COMMENTED** [review]
The idea certainly seems a good one, but wouldn't it be better to modify GSBlocks.h so that the CALL_BLOCK and CALL_BLOCK_NO_ARGS macros performed the check to see if the block is NULL?

**Comment by levinli303** [comment]
> The idea certainly seems a good one, but wouldn't it be better to modify GSBlocks.h so that the CALL_BLOCK and CALL_BLOCK_NO_ARGS macros performed the check to see if the block is NULL?

yeah I was thinking the same, if its acceptable, then I'd change this PR to modify the macros instead.

**Comment by rfm** [comment]
Yes, I think in the case where we have 'real' blocks, the compiler/runtime will check for a null block, but in the case where we are emulating blocks, we need the macro to add the code to do that check.

**Comment by levinli303** [comment]
@rfm on cocoa, calling a nil block would also crash.

**Comment by rfm** [comment]
Thanks ... good to know.  We should check for null in all versions of the macro then.

**Comment by levinli303** [comment]
@rfm then let's go ahead with this PR as is?

**Comment by rfm** [comment]
I expressed myself poorly.
I was agreeing with your earlier suggestion 
> I'd change this PR to modify the macros instead.

**Comment by levinli303** [comment]
@rfm ok, I've renamed the original macro that does not check nullability to CALL_NON_NULL_BLOCK, and use the original name for the case where nullability is checked. cannot use the checked one for cases where a return value is needed

File: Headers/GNUstepBase/GSBlocks.h [hunk 1]
```diff
@@ -100,13 +100,23 @@ typedef retTy(^name)()
...
+#define CALL_NON_NULL_BLOCK(block, args...) block->invoke(block, args)
+#define CALL_NON_NULL_BLOCK_NO_ARGS(block) block->invoke(block)
 #define BLOCK_SCOPE
 #endif /* GCC_VERSION >= 3000 */
 
 #endif /* __has_feature(blocks) */
 
+#define CALL_BLOCK(block, args...) do { \
+  if (NULL != block) \
+    CALL_NON_NULL_BLOCK(block, args); \
```
Comment (fredkiefer) [inline Headers/GNUstepBase/GSBlocks.h:112]: You just need a "return", plus an else case that will return "nil". This is not perfect, but better than this.

Comment (levinli303) [reply Headers/GNUstepBase/GSBlocks.h:112]: do you suggest that I change it to `#define CALL_BLOCK(block, args...) ((NULL != block) ? CALL_NON_NULL_BLOCK(block, args) : nil);` instead ?

Comment (rfm) [reply Headers/GNUstepBase/GSBlocks.h:112]: That looks good to me; clear and concise.

Comment (levinli303) [reply Headers/GNUstepBase/GSBlocks.h]: updated

Comment (levinli303) [reply Headers/GNUstepBase/GSBlocks.h]: @rfm  can this be merged?

**Pull request merged**